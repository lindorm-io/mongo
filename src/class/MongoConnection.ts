import { Db, MongoClient } from "mongodb";
import { IMongoConnectionOptions } from "../typing";

export class MongoConnection {
  private options: IMongoConnectionOptions;
  private client: MongoClient;
  private database: Db;

  constructor(options: IMongoConnectionOptions) {
    this.options = options;
  }

  private getUrl(): string {
    const { user, password, host, port } = this.options;
    return `mongodb://${user}:${password}@${host}:${port}/`;
  }

  public async connect(): Promise<void> {
    this.client = await MongoClient.connect(this.getUrl(), {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
    this.database = await this.client.db(this.options.name);
  }

  public db(): Db {
    if (!this.database) {
      throw new Error("You must connect() before you can get db()");
    }
    return this.database;
  }

  public async disconnect(): Promise<void> {
    await this.client.close();
  }
}
