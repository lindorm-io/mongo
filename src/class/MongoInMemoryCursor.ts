import { TObject } from "@lindorm-io/core";

export class MongoInMemoryCursor {
  public data: Array<TObject<any>>;

  constructor(data: Array<TObject<any>>) {
    this.data = data;
  }

  public async toArray(): Promise<Array<TObject<any>>> {
    return Promise.resolve(this.data);
  }
}
