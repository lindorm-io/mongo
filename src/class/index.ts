export * from "./MockMongoConnection";
export * from "./MongoConnection";
