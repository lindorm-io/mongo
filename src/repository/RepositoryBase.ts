import Joi from "@hapi/joi";
import { EntityCreationError, IEntity, TObject, TPromise } from "@lindorm-io/core";
import { Logger } from "@lindorm-io/winston";
import { RepositoryEntityNotFoundError, RepositoryEntityNotUpdatedError } from "../error";
import { TMongoCollection, TMongoDatabase } from "../typing";

export interface IRepository<Entity> {
  create(entity: Entity): Promise<Entity>;
  update(entity: Entity): Promise<Entity>;
  find(filter: TObject<any>): Promise<Entity>;
  findMany(filter: TObject<any>): Promise<Array<Entity>>;
  findOrCreate(filter: TObject<any>): Promise<Entity>;
  remove(entity: Entity): Promise<void>;
  removeMany(filter: TObject<any>): Promise<void>;
}

export interface IRepositoryOptions {
  db: TMongoDatabase;
  logger: Logger;
}

export interface IRepositoryBaseOptions extends IRepositoryOptions {
  collectionName: string;
  indices: Array<IIndex>;
  schema: Joi.Schema;
}

export interface IIndex {
  index: TObject<number>;
  options: {
    unique: boolean;
  };
}

export abstract class RepositoryBase<Entity extends IEntity> implements IRepository<Entity> {
  private collectionName: string;
  private collection: TMongoCollection;
  private db: TMongoDatabase;
  private indices: Array<IIndex>;
  protected logger: Logger;
  private promise: TPromise<void>;
  private schema: Joi.Schema;

  protected constructor(options: IRepositoryBaseOptions) {
    this.collectionName = options.collectionName;
    this.db = options.db;
    this.indices = options.indices;
    this.schema = options.schema;

    this.logger = options.logger.createChildLogger([
      "mongo",
      this.db.databaseName.toLowerCase(),
      this.collectionName.toLowerCase(),
    ]);
    this.promise = this.setup;
  }

  private async setup(): Promise<void> {
    const start = Date.now();
    this.collection = await this.db.collection(this.collectionName);

    for (const { index, options } of this.indices) {
      await this.collection.createIndex(index, options);
    }

    this.logger.debug("setup", {
      result: { success: true },
      time: Date.now() - start,
    });

    this.promise = () => Promise.resolve(undefined);
  }

  protected abstract createEntity(data: IEntity): Entity;

  protected abstract getEntityJSON(entity: Entity): IEntity;

  async create(entity: Entity): Promise<Entity> {
    const start = Date.now();

    try {
      entity.create();
    } catch (err) {
      if (!(err instanceof EntityCreationError)) {
        throw err;
      }
    }

    const json = this.getEntityJSON(entity);

    await this.schema.validateAsync(json);
    await this.promise();

    const result = await this.collection.insertOne(json);

    this.logger.debug("insertOne", {
      payload: Object.keys(json),
      result: { success: !!result.result.ok },
      time: Date.now() - start,
    });

    return entity;
  }

  async update(entity: Entity): Promise<Entity> {
    const start = Date.now();

    const currentVersion = entity.version;

    entity.updated = new Date();
    entity.version = currentVersion + 1;

    const json = this.getEntityJSON(entity);
    const { id, ...payload } = json;
    const filter = {
      id,
      version: { $eq: currentVersion },
    };

    await this.schema.validateAsync(json);
    await this.promise();

    const result = await this.collection.findOneAndUpdate(filter, { $set: payload });
    const success =
      result.ok && result.value && result.lastErrorObject?.n === 1 && result.lastErrorObject?.updatedExisting;

    this.logger.debug("findOneAndUpdate", {
      filter: Object.keys({ id }),
      payload: Object.keys(payload),
      result: { success },
      time: Date.now() - start,
    });

    if (!success) {
      throw new RepositoryEntityNotUpdatedError(filter, result);
    }

    return entity;
  }

  async find(filter: TObject<any>): Promise<Entity> {
    const start = Date.now();

    await this.promise();

    const result = await this.collection.findOne(filter);

    this.logger.debug("findOne", {
      filter: Object.keys(filter),
      result: { success: !!result },
      time: Date.now() - start,
    });

    if (!result) {
      throw new RepositoryEntityNotFoundError(filter, result);
    }

    return this.createEntity(result);
  }

  async findMany(filter: TObject<any> = {}): Promise<Array<Entity>> {
    const start = Date.now();

    await this.promise();

    const cursor = await this.collection.find(filter);
    const results = await cursor.toArray();

    this.logger.debug("find", {
      filter: Object.keys(filter),
      result: { success: !!results.length, amount: results.length },
      time: Date.now() - start,
    });

    const entities: Array<Entity> = [];

    for (const item of results) {
      entities.push(this.createEntity(item));
    }

    return entities;
  }

  async findOrCreate(filter: TObject<any>): Promise<Entity> {
    try {
      return await this.find(filter);
    } catch (err) {
      if (err instanceof RepositoryEntityNotFoundError) {
        return await this.create(this.createEntity(filter as IEntity));
      }
      throw err;
    }
  }

  async remove(entity: Entity): Promise<void> {
    const start = Date.now();

    const { id } = this.getEntityJSON(entity);

    await this.promise();

    const result = await this.collection.findOneAndDelete({ id });

    this.logger.debug("findOneAndDelete", {
      filter: Object.keys({ id }),
      result: { success: !!result.ok },
      time: Date.now() - start,
    });
  }

  async removeMany(filter: TObject<any>): Promise<void> {
    const start = Date.now();

    await this.promise();

    const result = await this.collection.deleteMany(filter);

    this.logger.debug("deleteMany", {
      filter: Object.keys(filter),
      result: { success: !!result.result.ok, amount: result.result.n },
      time: Date.now() - start,
    });
  }
}
