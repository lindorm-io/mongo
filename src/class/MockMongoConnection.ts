import lodash from "lodash";
import { TObject } from "@lindorm-io/core";
import { IMongoConnectionOptions } from "./MongoConnection";

export class MockMongoCursor {
  public data: Array<TObject<any>>;

  constructor(data: Array<TObject<any>>) {
    this.data = data;
  }

  public toArray(): Promise<Array<TObject<any>>> {
    return Promise.resolve(this.data);
  }
}

export class MockMongoCollection {
  public data: Array<TObject<any>>;
  public indices: Array<TObject<any>>;

  constructor() {
    this.indices = [];
    this.data = [];
  }

  public createIndex(index: TObject<any>, options: TObject<any>): Promise<any> {
    this.indices.push({ index, options });
    return Promise.resolve();
  }

  public insertOne(json: TObject<any>): Promise<any> {
    this.data.push(json);
    return Promise.resolve({
      result: {
        ok: true,
      },
    });
  }

  public findOneAndUpdate(filter: TObject<any>, options: TObject<any>): Promise<any> {
    const {
      version: { $eq },
      ...params
    } = filter;
    const query = { ...params, version: $eq };
    const item = lodash.find(this.data, query);
    if (!item) {
      return Promise.resolve({
        ok: false,
        value: 0,
        lastErrorObject: {
          n: 0,
          updatedExisting: 0,
        },
      });
    }
    lodash.remove(this.data, query);
    this.data.push({ ...params, ...options.$set });
    return Promise.resolve({
      ok: true,
      value: 1,
      lastErrorObject: {
        n: 1,
        updatedExisting: 1,
      },
    });
  }

  public findOne(filter: TObject<any>): Promise<any> {
    return Promise.resolve(lodash.find(this.data, filter));
  }

  public find(filter: TObject<any>): Promise<MockMongoCursor> {
    return Promise.resolve(new MockMongoCursor(lodash.filter(this.data, filter)));
  }

  public findOneAndDelete(filter: TObject<any>): Promise<any> {
    lodash.remove(this.data, filter);
    return Promise.resolve({
      ok: true,
    });
  }

  public deleteMany(filter: TObject<any>): Promise<any> {
    const array = lodash.filter(this.data, filter);
    lodash.remove(this.data, filter);
    return Promise.resolve({
      result: {
        ok: true,
        n: array.length,
      },
    });
  }
}

export class MockMongoDatabase {
  public collections: TObject<MockMongoCollection>;
  public databaseName: string;

  constructor(name: string) {
    this.collections = {};
    this.databaseName = name;
  }

  public collection(name: string): Promise<MockMongoCollection> {
    if (!this.collections[name]) {
      this.collections[name] = new MockMongoCollection();
    }
    return Promise.resolve(this.collections[name]);
  }
}

export class MockMongoClient {
  public databases: TObject<MockMongoDatabase>;

  constructor() {
    this.databases = {};
  }

  public db(name: string): Promise<MockMongoDatabase> {
    if (!this.databases[name]) {
      this.databases[name] = new MockMongoDatabase(name);
    }
    return Promise.resolve(this.databases[name]);
  }

  public close(): Promise<void> {
    return Promise.resolve();
  }
}

export class MockMongo {
  public client: MockMongoClient;

  constructor(_: IMongoConnectionOptions) {
    this.client = new MockMongoClient();
  }

  public connect(): Promise<MockMongoClient> {
    return Promise.resolve(this.client);
  }
}
