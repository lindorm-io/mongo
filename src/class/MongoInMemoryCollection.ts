import lodash from "lodash";
import { MongoInMemoryCursor } from "./MongoInMemoryCursor";
import { TObject } from "@lindorm-io/core";

export class MongoInMemoryCollection {
  public collection: TObject<any>;

  constructor(collection: TObject<any>) {
    this.collection = collection;

    if (!this.collection.indices) {
      this.collection.indices = [];
    }
    if (!this.collection.data) {
      this.collection.data = [];
    }
  }

  public async createIndex(index: TObject<any>, options: TObject<any>): Promise<any> {
    this.collection.indices.push({ index, options });

    return Promise.resolve();
  }

  public async insertOne(json: TObject<any>): Promise<any> {
    this.collection.data.push(json);

    return Promise.resolve({
      result: {
        ok: true,
      },
    });
  }

  public async findOneAndUpdate(filter: TObject<any>, options: TObject<any>): Promise<any> {
    const {
      version: { $eq },
      ...params
    } = filter;

    const query = { ...params, version: $eq };
    const item = lodash.find(this.collection.data, query);

    if (!item) {
      return Promise.resolve({
        ok: false,
        value: 0,
        lastErrorObject: {
          n: 0,
          updatedExisting: 0,
        },
      });
    }

    lodash.remove(this.collection.data, query);

    this.collection.data.push({ ...params, ...options.$set });

    return Promise.resolve({
      ok: true,
      value: 1,
      lastErrorObject: {
        n: 1,
        updatedExisting: 1,
      },
    });
  }

  public async findOne(filter: TObject<any>): Promise<any> {
    return Promise.resolve(lodash.find(this.collection.data, filter));
  }

  public async find(filter: TObject<any>): Promise<MongoInMemoryCursor> {
    return Promise.resolve(new MongoInMemoryCursor(lodash.filter(this.collection.data, filter)));
  }

  public async findOneAndDelete(filter: TObject<any>): Promise<any> {
    lodash.remove(this.collection.data, filter);

    return Promise.resolve({
      ok: true,
    });
  }

  public async deleteMany(filter: TObject<any>): Promise<any> {
    const array = lodash.filter(this.collection.data, filter);

    lodash.remove(this.collection.data, filter);

    return Promise.resolve({
      result: {
        ok: true,
        n: array.length,
      },
    });
  }
}
