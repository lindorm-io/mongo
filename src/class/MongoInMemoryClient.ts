import { MongoInMemoryDatabase } from "./MongoInMemoryDatabase";
import { TObject } from "@lindorm-io/core";

export const MONGO_IN_MEMORY_DB: TObject<any> = {};

export class MongoInMemoryClient {
  public databases: TObject<MongoInMemoryDatabase>;

  constructor() {
    this.databases = {};
  }

  public async db(databaseName: string): Promise<MongoInMemoryDatabase> {
    if (!this.databases[databaseName]) {
      MONGO_IN_MEMORY_DB[databaseName] = {};
    }

    this.databases[databaseName] = new MongoInMemoryDatabase(databaseName);

    return this.databases[databaseName];
  }

  public async close(): Promise<void> {
    this.databases = {};
    return Promise.resolve();
  }
}
