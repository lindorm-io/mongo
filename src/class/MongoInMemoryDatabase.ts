import { MONGO_IN_MEMORY_DB } from "./MongoInMemoryClient";
import { MongoInMemoryCollection } from "./MongoInMemoryCollection";
import { TObject } from "@lindorm-io/core";

export class MongoInMemoryDatabase {
  public collections: TObject<MongoInMemoryCollection>;
  public databaseName: string;

  constructor(databaseName: string) {
    this.collections = {};
    this.databaseName = databaseName;
  }

  public async collection(collectionName: string): Promise<MongoInMemoryCollection> {
    if (!this.collections[collectionName]) {
      MONGO_IN_MEMORY_DB[this.databaseName][collectionName] = {};
    }

    this.collections[collectionName] = new MongoInMemoryCollection(this.databaseName, collectionName);

    return Promise.resolve(this.collections[collectionName]);
  }
}
