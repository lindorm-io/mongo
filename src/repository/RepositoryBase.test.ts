import Joi from "@hapi/joi";
import { IRepositoryOptions, RepositoryBase } from "./RepositoryBase";
import { EntityBase } from "@lindorm-io/core";
import { MockMongo, MockMongoDatabase } from "../class";

class MockEntity extends EntityBase {
  public name: string;

  constructor(options: any) {
    super(options);
    this.name = options.name;
  }
}

class MockRepository extends RepositoryBase<MockEntity> {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: "MockRepository",
      db: options.db,
      logger: options.logger,
      indices: [{ index: { id: 1 }, options: { unique: true } }],
      schema: Joi.object({
        created: Joi.date().required(),
        id: Joi.string().guid().required(),
        name: Joi.string().required(),
        updated: Joi.date().required(),
        version: Joi.number().required(),
      }),
    });
  }

  protected createEntity(data: any): MockEntity {
    return new MockEntity(data);
  }

  protected getEntityJSON(entity: MockEntity): any {
    return {
      created: entity.created,
      id: entity.id,
      name: entity.name,
      updated: entity.updated,
      version: entity.version,
    };
  }
}

describe("RepositoryBase.ts", () => {
  let database: MockMongoDatabase;
  let repository: MockRepository;

  beforeEach(async () => {
    const mongo = new MockMongo({ log: jest.fn() });
    const client = await mongo.connect("client", { mock: true });

    database = await client.db("database");
    repository = new MockRepository({
      // @ts-ignore
      db: database,
      logger: {
        // @ts-ignore
        createChildLogger: jest.fn(() => ({
          debug: jest.fn(),
        })),
      },
    });
  });

  test("should run setup on any method", async () => {
    await repository.create(new MockEntity({ name: "mock" }));
    const collection = database.collections["MockRepository"];

    expect(collection.indices).toStrictEqual([
      {
        index: {
          id: 1,
        },
        options: {
          unique: true,
        },
      },
    ]);
  });

  test("should create entity", async () => {
    const e1 = new MockEntity({ name: "e1" });
    const created = await repository.create(e1);

    expect(created).toStrictEqual(e1);
  });

  test("should update entity", async () => {
    const e1 = new MockEntity({ name: "e1" });
    const created = await repository.create(e1);

    created.name = "changed";

    const updated = await repository.update(created);

    expect(updated).toStrictEqual(created);
  });

  test("should find entity", async () => {
    const e1 = new MockEntity({ name: "e1" });
    await repository.create(e1);

    const found = await repository.find({ name: "e1" });

    expect(found).toStrictEqual(e1);
  });

  test("should find many entities", async () => {
    const e1 = new MockEntity({ name: "ex" });
    const e2 = new MockEntity({ name: "ex" });

    await repository.create(e1);
    await repository.create(e2);

    const found = await repository.findMany({ name: "ex" });

    expect(found).toStrictEqual([e1, e2]);
  });

  test("should remove one entity", async () => {
    const e1 = new MockEntity({ name: "e1" });
    const e2 = new MockEntity({ name: "e2" });

    await repository.create(e1);
    await repository.create(e2);

    await repository.remove(e1);

    await expect(repository.findMany({})).resolves.toStrictEqual([e2]);
  });

  test("should remove many entities", async () => {
    const e1 = new MockEntity({ name: "ex" });
    const e2 = new MockEntity({ name: "ex" });
    const e3 = new MockEntity({ name: "e3" });

    await repository.create(e1);
    await repository.create(e2);
    await repository.create(e3);

    await repository.removeMany({ name: "ex" });

    await expect(repository.findMany({})).resolves.toStrictEqual([e3]);
  });
});
