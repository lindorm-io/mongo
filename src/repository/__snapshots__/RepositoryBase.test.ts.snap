// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryBase.ts should create entity 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "name": "e1",
}
`;

exports[`RepositoryBase.ts should find entity 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "name": "e1",
}
`;

exports[`RepositoryBase.ts should find many entities 1`] = `
Array [
  MockEntity {
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
    "name": "ex",
  },
  MockEntity {
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
    "name": "ex",
  },
]
`;

exports[`RepositoryBase.ts should remove many entities 1`] = `
Array [
  MockEntity {
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
    "name": "e3",
  },
]
`;

exports[`RepositoryBase.ts should remove one entity 1`] = `Array []`;

exports[`RepositoryBase.ts should run setup on any method 1`] = `
Array [
  Object {
    "index": Object {
      "id": 1,
    },
    "options": Object {
      "unique": true,
    },
  },
]
`;

exports[`RepositoryBase.ts should update entity 1`] = `
MockEntity {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "e397bc49-849e-4df6-a536-7b9fa3574ace",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 1,
  "name": "changed",
}
`;
